### PR TITLE
Polylang: Fail if we download the incorrect version

### PR DESCRIPTION
**Pull request by jarstelfox**
We are seeing the api give us the most recent version, but this version broke our website.

We spent a lot of time running down the fact that the version we requested was not being downloaded.

https://polylang.pro?edd_action=get_version&license=REDACTED&item_name=Polylang+Pro&url=REDACTED&version=3.1.4

returns

{
    "new_version": "3.2.3",
    "stable_version": "3.2.3",
    "name": "Polylang Pro",
    "slug": "polylang-pro",
    ...
}

It even says it is installing 3.1.4 during the composer install step.

I would have liked it to fail loudly when this occurs.

### CODE REVIEW

**Review by djmetzle — APPROVED** [review]

**Comment by mcaskill** [comment]
This is a good idea. Would prevent some headaches I've encountered in the past.
Can be applied to other plugins using Easy Digital Downloads.

File: plugins/PolylangPro.php [hunk 1]
```diff
@@ -44,6 +44,14 @@ public function getDownloadUrl() {
 			'url'        => getenv( 'POLYLANG_PRO_URL' ),
 			'version'    => $this->version,
 		) ), true );
+
+		/*
+		 * If the response is not the version we asked for then, bail
+		 */
+		if ( $this->version != $response['stable_version']) ) {
```
Comment (junaidbhura) [inline plugins/PolylangPro.php]: @jarstelfox Could you please make this:

`if ( $response['stable_version'] !== $this->version  ) ) {` to be in line with WP coding standards?

**Review by junaidbhura — CHANGES_REQUESTED** [review]

**Comment by junaidbhura** [comment]
@mcaskill Thanks, could you elaborate what problems this could fix?

**Comment by mcaskill** [comment]
> […] could you elaborate what problems this could fix?

For example: say your project requires Polylang 3.3.0. Since Polylang's EDD only returns the latest version, by aborting when there's a version conflict, you prevent Composer from caching this new version (3.4.0) under the key constructed from the required/locked version (3.3.0). This saves you from having to delete the cached download _and_ alerts you to the existance of an update (which could be breaking).

**Comment by junaidbhura** [comment]
I see thanks @mcaskill . @jarstelfox happy to merge this in once you make that update!

**Comment by mcaskill** [comment]
It should be noted that I have not tested this pull request yet.

File: plugins/PolylangPro.php (hunk 1 above)
Comment (jarstelfox) [reply plugins/PolylangPro.php]:
```suggestion
		if ( $response['stable_version'] !== $this->version ) ) {
```

**Comment by mcaskill** [comment]
Should the equivalent be applied to WP All Import / Export Pro (`WpAiPro`) or should that be a separate pull request?

File: plugins/PolylangPro.php [hunk 2]
```diff
@@ -44,6 +44,14 @@ public function getDownloadUrl() {
 			'url'        => getenv( 'POLYLANG_PRO_URL' ),
 			'version'    => $this->version,
 		) ), true );
+
+		/*
+		 * If the response is not the version we asked for then, bail
+		 */
+		if ( $response['stable_version'] !== $this->version ) ) {
+			return '';
+		}
```
Comment (mcaskill) [inline plugins/PolylangPro.php:53]: Fix syntax error (extra `)`), replace `stable_version` with `new_version` according to EDD SL API, and ensure `new_version` exists.

```suggestion
		/**
		 * If the response does not have a version number or the version number
		 * does not match the package constraint, bail.
		 */
		if ( empty( $response['new_version'] ) || $response['new_version'] !== $this->version ) {
			return '';
		}
```

**Review by mcaskill — CHANGES_REQUESTED** [review]
I was doing a bit of reading on EDD's [Software Licensing API](https://easydigitaldownloads.com/docs/software-licensing-api/) as well as its [version history](https://easydigitaldownloads.com/downloads/software-licensing/).

The API response includes two version attributes, `new_version` and `stable_version`.

Since SL 3.5.0 (2017-01-13), [beta versions](https://easydigitaldownloads.com/docs/software-licensing-releasing-beta-versions/) are supported by appending `&beta=1` to the request. If a beta version is available, the `new_version` attribute will contain that value and the `download_link` will be for the beta.

Since SL 3.5.4 (2017-01-30), the `stable_version` attribute was introduced. Most likely to help distinguish between beta and stable versions (akin to GitHub's "Latest Release" option).

Since SL 3.8.0 (2021-08-30), you can include platform versions in the request, by appending `&php_version=<version>` and `&wp_version=<version>`. If the platform version does not meet the minimum requirements set for the product in EDD, the `new_version` in the response will be `false` instead of the new version number. This is an interesting feature that can be used to bail if incompatible (for a future PR).

To summarize, the `download_link` attribute will be either for the beta or the stable, which means corresponding to the `new_version` attribute.

**Comment by mcaskill** [comment]
I've been doing some tests and I've stumbled upon an edge case with the current logic.

Say the plugin's latest version is `3.3` but your package constraint is `3.3.0`, the identical operator will return `false`.

The solution to that is to use [composer/semver](https://github.com/composer/semver):

```php
use Composer\Semver\Semver;

if ( empty( $response['download_link'] ) ) {
	return '';
}

if ( empty( $response['new_version'] ) ) {
	return '';
}

if ( ! Semver::satisfies( $response['new_version'], $this->version ) ) {
	return '';
}

return $response['download_link'];
```

I'm using `Semver` class since it has the widest feature support compared to Composer's `Comparator::equalTo()` or PHP's `version_compare()` which does not consider `1` and `1.0` and `1.0.0` as equal.

The composer/semver package is required by composer/composer but since junaidbhura/composer-wp-pro-plugins would be directly using it, we would have to add that package to the plugin's requirements.

**Comment by junaidbhura** [comment]
@mcaskill @jarstelfox Closing this PR in favor of #47
